# Introduction to Pandas

In this tutorial, we will learn how to use Pandas by analyzing a real-world dataset.

The dataset that we are going to analyze is the TED talk dataset which is available on Kaggle (https://www.kaggle.com/datasets/ahmadfatani/ted-talks-dataset). The dataset contains information about all video recordings of TED Talks uploaded to the official TED.com website until April 18th, 2020. It contains information about all talks including the number of views, tags, posted-date, speakers and titles.

Note that you do not have to download the dataset from Kaggle since the data is already contained in the Github repository.

## Data Exploration

As a first step, we want to get a basic understanding of the data contained in the dataset. In particular, we want to know things such as:

- What data (e.g., columns) is available?
- How large is our dataset (how many rows are there)?
- What datatype does each column have?
- Get some statistics on our data

### Loading the CSV file

The first step is to load the data provided in the csv file. The file that we want to load is `ted_main.csv`.

With the `pd.read_csv()` method, Pandas provides a convenient way to read csv files.

In [1]:
import pandas as pd

In [2]:
# By default, ted_main.csv expects the csv file to use "," as a separator and no delimiters.
# Of course, this can be changed by setting the <sep> and <delimiter> parameters accordingly.
ted_df = pd.read_csv('../ted_talk_dataset/ted_main.csv')

### Printing the first few rows

A good way to start inspecting a dataset is to look at the first few rows. This can be easily done using the `head()` method. <br/>
By default, `head()` prints the first 5 rows of a dataframe

In [3]:
ted_df.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60   Ken Robinson   
1  TED2006  1140825600         43        Al Gore   
2  TED2006  1140739200         26    David Pogue   
3  TED2006  1140912000         35  Majora Carter   
4  TED2006  1140566400         48   Hans Rosling   

                                            name  num_speaker  published_date  \
0      Ken Robinson: Do schools kill creativity?            1      1151367060   
1           Al Gore: Averting the climate crisis            1      1151367060   
2                  David Pogue: Simplicity sells            1      1151367060   
3             Majora Carter: Greening the ghetto            1      1151367060   
4  Hans Rosling: The best stats you've ever seen            1      1151440680   

                                             ratings  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   

                                       related_talks  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520  
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292  
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550  
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869

We can also change the number of rows we want to print. Let's say we want to look at the first two rows only.

In [4]:
ted_df.head(2)

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   

     event   film_date  languages  main_speaker  \
0  TED2006  1140825600         60  Ken Robinson   
1  TED2006  1140825600         43       Al Gore   

                                        name  num_speaker  published_date  \
0  Ken Robinson: Do schools kill creativity?            1      1151367060   
1       Al Gore: Averting the climate crisis            1      1151367060   

                                             ratings  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   

                                       related_talks speaker_occupation  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...    Author/educator   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   Climate advocate   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   

                         title  \
0  Do schools kill creativity?   
1  Averting the climate crisis   

                                                 url     views  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520

Alternatively, we can also inspect the end of the list using `tail()`.

In [5]:
ted_df.tail(2)

comments                                        description  duration  \
2548        32  In an unmissable talk about race and politics ...      1100   
2549         8  With more than half of the world population li...       519   

             event   film_date  languages       main_speaker  \
2548  TEDxMileHigh  1499472000          1   Theo E.J. Wilson   
2549       TED2017  1492992000          1  Karoliina Korppoo   

                                                   name  num_speaker  \
2548  Theo E.J. Wilson: A black man goes undercover ...            1   
2549  Karoliina Korppoo: How a video game might help...            1   

      published_date                                            ratings  \
2548      1506024042  [{'id': 11, 'name': 'Longwinded', 'count': 3},...   
2549      1506092422  [{'id': 21, 'name': 'Unconvincing', 'count': 2...   

                                          related_talks   speaker_occupation  \
2548  [{'id': 2512, 'hero': 'https://pe.tedcdn.com/i...  Public intellectual   
2549  [{'id': 2682, 'hero': 'https://pe.tedcdn.com/i...        Game designer   

                                                   tags  \
2548  ['Internet', 'TEDx', 'United States', 'communi...   
2549  ['cities', 'design', 'future', 'infrastructure...   

                                                  title  \
2548       A black man goes undercover in the alt-right   
2549  How a video game might help us build better ci...   

                                                    url   views  
2548  https://www.ted.com/talks/theo_e_j_wilson_a_bl...  419309  
2549  https://www.ted.com/talks/karoliina_korppoo_ho...  391721

### Obtaining the shape of a dataframe

When we look at the first two rows, we can immediately see that the dataframe contains multiple columns. But how many columns and rows does it have?

We can inspect the size of a dataframe by writing ...

In [6]:
# Returns df shape as tuple rows x columns
print(ted_df.shape)

print('Num rows:', ted_df.shape[0])
print('Num cols:', ted_df.shape[1])

(2550, 17)
Num rows: 2550
Num cols: 17


### Inspecting each column's data type

Next, take a look at what Pandas thinks of the data types by checking the attribute `dtype`.

In [7]:
ted_df.dtypes

comments               int64
description           object
duration               int64
event                 object
film_date              int64
languages              int64
main_speaker          object
name                  object
num_speaker            int64
published_date         int64
ratings               object
related_talks         object
speaker_occupation    object
tags                  object
title                 object
url                   object
views                  int64
dtype: object

Ok, that's interesting. Apparently, Pandas considers many columns as data type `object` (not `str`). So, what's going on here? <br/>
Pandas stores data (values) in a contiguous, fixed-size memory block. For int64 and float64, they are 8 bytes. But for strings, the length of the string is not fixed. To overcome this problem, pandas stores the data in an object array where each element simply contains the address of a string.

A list of available dtypes can be found here: https://pandas.pydata.org/docs/user_guide/basics.html#basics-dtypes

In [8]:
print(ted_df.dtypes)

comments               int64
description           object
duration               int64
event                 object
film_date              int64
languages              int64
main_speaker          object
name                  object
num_speaker            int64
published_date         int64
ratings               object
related_talks         object
speaker_occupation    object
tags                  object
title                 object
url                   object
views                  int64
dtype: object


Another option to obtain similar information from a dataframe with just one glance is the `info()` method. This method provides a concise summary of a dataframe.

In [9]:
ted_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2550 entries, 0 to 2549
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   comments            2550 non-null   int64 
 1   description         2550 non-null   object
 2   duration            2550 non-null   int64 
 3   event               2550 non-null   object
 4   film_date           2550 non-null   int64 
 5   languages           2550 non-null   int64 
 6   main_speaker        2550 non-null   object
 7   name                2550 non-null   object
 8   num_speaker         2550 non-null   int64 
 9   published_date      2550 non-null   int64 
 10  ratings             2550 non-null   object
 11  related_talks       2550 non-null   object
 12  speaker_occupation  2544 non-null   object
 13  tags                2550 non-null   object
 14  title               2550 non-null   object
 15  url                 2550 non-null   object
 16  views               2550

As can be seen, the information provided by `info()` also tells us whether a column contains cells without any value. <br/>
Indeed, this seems to be the case for the column `speaker_occupation` where six rows do not contain any value.

### Filtering empty rows in a dataframe

Sometimes we might want to remove empty rows from a dataframe. The following lines illustrate how this can be achieved.

First, we look for cells that contain no values with the `isna()` method. The output is a binary mask.

In [10]:
ted_df.isna()

comments  description  duration  event  film_date  languages  \
0        False        False     False  False      False      False   
1        False        False     False  False      False      False   
2        False        False     False  False      False      False   
3        False        False     False  False      False      False   
4        False        False     False  False      False      False   
...        ...          ...       ...    ...        ...        ...   
2545     False        False     False  False      False      False   
2546     False        False     False  False      False      False   
2547     False        False     False  False      False      False   
2548     False        False     False  False      False      False   
2549     False        False     False  False      False      False   

      main_speaker   name  num_speaker  published_date  ratings  \
0            False  False        False           False    False   
1            False  False        False           False    False   
2            False  False        False           False    False   
3            False  False        False           False    False   
4            False  False        False           False    False   
...            ...    ...          ...             ...      ...   
2545         False  False        False           False    False   
2546         False  False        False           False    False   
2547         False  False        False           False    False   
2548         False  False        False           False    False   
2549         False  False        False           False    False   

      related_talks  speaker_occupation   tags  title    url  views  
0             False               False  False  False  False  False  
1             False               False  False  False  False  False  
2             False               False  False  False  False  False  
3             False               False  False  False  False  False  
4             False               False  False  False  False  False  
...             ...                 ...    ...    ...    ...    ...  
2545          False               False  False  False  False  False  
2546          False               False  False  False  False  False  
2547          False               False  False  False  False  False  
2548          False               False  False  False  False  False  
2549          False               False  False  False  False  False  

[2550 rows x 17 columns]

We can now easily count the number of `True`s in each column using `sum()`. <br/>
By default, `sum()` computes the sum along the columns. `True` is regarded as one. `False` is regarded as 0.

In [11]:
ted_df.isna().sum()

comments              0
description           0
duration              0
event                 0
film_date             0
languages             0
main_speaker          0
name                  0
num_speaker           0
published_date        0
ratings               0
related_talks         0
speaker_occupation    6
tags                  0
title                 0
url                   0
views                 0
dtype: int64

If we really want to remove the non-zero rows, we can do this by writing ...

In [12]:
# Remove all nan values from the dataframe
ted_df.dropna(inplace=True)

In [13]:
# New let' check again to see whether the NaNs are gone
print(ted_df.isna().sum())

comments              0
description           0
duration              0
event                 0
film_date             0
languages             0
main_speaker          0
name                  0
num_speaker           0
published_date        0
ratings               0
related_talks         0
speaker_occupation    0
tags                  0
title                 0
url                   0
views                 0
dtype: int64


### Obtaining some overall descriptive statistics

All pandas dataframes provide a `describe()` method that allows you to easily obtain some descriptive statistics from the data. It provides information about the central tendency, dispersion, and shape of a dataset's distribution, excluding `NaN` values (object columns are not analyzed).

By default, only numeric columns are included.

In [14]:
ted_df.describe()

comments     duration     film_date    languages  num_speaker  \
count  2544.000000  2544.000000  2.544000e+03  2544.000000  2544.000000   
mean    191.706761   827.316431  1.321828e+09    27.319969     1.028302   
std     282.613719   373.828955  1.198455e+08     9.563529     0.207945   
min       2.000000   135.000000  7.464960e+07     0.000000     1.000000   
25%      63.000000   578.750000  1.257466e+09    23.000000     1.000000   
50%     118.000000   848.500000  1.333238e+09    28.000000     1.000000   
75%     222.000000  1047.000000  1.412921e+09    33.000000     1.000000   
max    6404.000000  5256.000000  1.503792e+09    72.000000     5.000000   

       published_date         views  
count    2.544000e+03  2.544000e+03  
mean     1.343456e+09  1.699779e+06  
std      9.471837e+07  2.501043e+06  
min      1.151367e+09  5.044300e+04  
25%      1.268341e+09  7.565802e+05  
50%      1.340935e+09  1.123870e+06  
75%      1.423519e+09  1.702149e+06  
max      1.506092e+09  4.722711e+07

As can be seen, 50% of all talks have around 118 comments. On average, they have 191 comments. The average duration of a TED talk is ~826 seconds (13-14 minutes) long. <br/>


### Obtaining information about the remaining columns

As mentioned previously, by default, `describe()` only considers columns with numeric values. However, we can also force `describe()` to look at other data types as well.

Let's force Python to only look at objects (which are strings in our case).

In [15]:
ted_df.describe(include=object)

description    event  \
count                                                2544     2544   
unique                                               2544      355   
top     Sir Ken Robinson makes an entertaining and pro...  TED2014   
freq                                                    1       84   

        main_speaker                                       name  \
count           2544                                       2544   
unique          2150                                       2544   
top     Hans Rosling  Ken Robinson: Do schools kill creativity?   
freq               9                                          1   

                                                  ratings  \
count                                                2544   
unique                                               2544   
top     [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
freq                                                    1   

                                            related_talks speaker_occupation  \
count                                                2544               2544   
unique                                               2544               1458   
top     [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...             Writer   
freq                                                    1                 45   

                         tags                        title  \
count                    2544                         2544   
unique                   2524                         2544   
top     ['art', 'creativity']  Do schools kill creativity?   
freq                        3                            1   

                                                      url  
count                                                2544  
unique                                               2544  
top     https://www.ted.com/talks/ken_robinson_says_sc...  
freq                                                    1

**Hint:** "Freq" denotes the most common value's frequency.

Alternatively, we can disable both types of columns at the same time.

In [16]:
ted_df.describe(include='all')

comments                                        description  \
count   2544.000000                                               2544   
unique          NaN                                               2544   
top             NaN  Sir Ken Robinson makes an entertaining and pro...   
freq            NaN                                                  1   
mean     191.706761                                                NaN   
std      282.613719                                                NaN   
min        2.000000                                                NaN   
25%       63.000000                                                NaN   
50%      118.000000                                                NaN   
75%      222.000000                                                NaN   
max     6404.000000                                                NaN   

           duration    event     film_date    languages  main_speaker  \
count   2544.000000     2544  2.544000e+03  2544.000000          2544   
unique          NaN      355           NaN          NaN          2150   
top             NaN  TED2014           NaN          NaN  Hans Rosling   
freq            NaN       84           NaN          NaN             9   
mean     827.316431      NaN  1.321828e+09    27.319969           NaN   
std      373.828955      NaN  1.198455e+08     9.563529           NaN   
min      135.000000      NaN  7.464960e+07     0.000000           NaN   
25%      578.750000      NaN  1.257466e+09    23.000000           NaN   
50%      848.500000      NaN  1.333238e+09    28.000000           NaN   
75%     1047.000000      NaN  1.412921e+09    33.000000           NaN   
max     5256.000000      NaN  1.503792e+09    72.000000           NaN   

                                             name  num_speaker  \
count                                        2544  2544.000000   
unique                                       2544          NaN   
top     Ken Robinson: Do schools kill creativity?          NaN   
freq                                            1          NaN   
mean                                          NaN     1.028302   
std                                           NaN     0.207945   
min                                           NaN     1.000000   
25%                                           NaN     1.000000   
50%                                           NaN     1.000000   
75%                                           NaN     1.000000   
max                                           NaN     5.000000   

        published_date                                            ratings  \
count     2.544000e+03                                               2544   
unique             NaN                                               2544   
top                NaN  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
freq               NaN                                                  1   
mean      1.343456e+09                                                NaN   
std       9.471837e+07                                                NaN   
min       1.151367e+09                                                NaN   
25%       1.268341e+09                                                NaN   
50%       1.340935e+09                                                NaN   
75%       1.423519e+09                                                NaN   
max       1.506092e+09                                                NaN   

                                            related_talks speaker_occupation  \
count                                                2544               2544   
unique                                               2544               1458   
top     [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...             Writer   
freq                                                    1                 45   
mean                                                  NaN                NaN   
std                                            